File: Algorithm/BruteForce/거리두기 확인하기.cpp
#include <string>
#include <vector>
#include <cstring>

using namespace std;

int flag = 1;
int dir[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
int visited[5][5];

void DFS(vector<string> place, int x, int y, int depth) {
    if (!flag) return;
    
    if (depth == 2) return;
    
    for (auto& d: dir)
    {
        int nx = x + d[0];
        int ny = y + d[1];
        
        if (nx < 0 || ny < 0 || nx >= 5 || ny >= 5) 
            continue;
        
        if (visited[nx][ny]) 
            continue;
        
        if (place[nx][ny] == 'X')
            continue;
        
        if (depth < 2 && place[nx][ny] == 'P')
        {
            flag = 0;
            return;
        }
        
        visited[nx][ny] = 1;
        DFS(place, nx, ny, depth+1);
    }
    
    return;
}

int isSafeRule(vector<string> place) {
    flag = 1;
    for (int i = 0; i < 5; ++i)
    {
        for (int j = 0; j < 5; ++j)
        {
            if (place[i][j] == 'P')
            {
                memset(visited, 0, sizeof(visited));
                visited[i][j] = true;
                DFS(place, i, j, 0);
            }
            
            if (!flag) break;
        }
    }
    return flag;
}

vector<int> solution(vector<vector<string>> places) {
    vector<int> answer;
    
    for (auto& place: places) 
    {
        answer.push_back(isSafeRule(place));
    }
    return answer;
}

File: Algorithm/BruteForce/거리두기 확인하기.py
flag = True
dir = [[1,0], [-1,0], [0,1], [0,-1]]

def solution(places):
    answer = []
    
    for place in places:
        answer.append(checkProtect(place))
    return answer

def checkProtect(place):
    global flag
    flag = True
    visited = [[False]*5 for _ in range(5)]
    
    for i in range(len(place)):
        for j in range(len(place[0])):
            if place[i][j] == "P":
                visited[i][j] = True
                DFS(place, visited, i, j, 0)
                
                if not flag:
                    return 0
    
    return 1

def DFS(place, visited, x, y, depth):
    global flag
    
    if not flag:
        return
    
    if flag and depth == 2:
        return
    
    for d in dir:
        nx = x + d[0]
        ny = y + d[1]
        
        if nx < 0 or ny < 0 or nx >= 5 or ny >= 5:
            continue
            
        if visited[nx][ny] or place[nx][ny] == "X":
            continue
            
        if place[nx][ny] == "P":
            flag = False
            return
        
        visited[nx][ny] = True
        DFS(place, visited, nx, ny, depth+1)
    return